cmd/xl-sets.go

```diff
@@ -322,38 +317,50 @@ func newXLSets(endpoints Endpoints, format *formatXLV3, setCount int, drivesPerS
 	// setCount * drivesPerSet with each memory upto blockSizeV1.
 	bp := bpool.NewBytePoolCap(setCount*drivesPerSet, blockSizeV1, blockSizeV1*2)
 
-	for i := 0; i < len(format.XL.Sets); i++ {
+	var wg sync.WaitGroup
+	var onlineDisks int
+	for i := 0; i < setCount; i++ {
 		s.xlDisks[i] = make([]StorageAPI, drivesPerSet)
 		s.xlLockers[i] = make([]dsync.NetLocker, drivesPerSet)
+		wg.Add(1)
+		go func(i int) {
+			defer wg.Done()
+			var endpoints Endpoints
+			for j := 0; j < drivesPerSet; j++ {
+				endpoints = append(endpoints, s.endpoints[i*drivesPerSet+j])
+				// Rely on endpoints list to initialize, init lockers and available disks.
+				s.xlLockers[i][j] = newLockAPI(s.endpoints[i*drivesPerSet+j])
+				s.xlDisks[i][j] = storageDisks[i*drivesPerSet+j]
+			}
 
-		var endpoints Endpoints
-		for j := 0; j < drivesPerSet; j++ {
-			endpoints = append(endpoints, s.endpoints[i*s.drivesPerSet+j])
-		}
+			// Initialize xl objects for a given set.
+			s.sets[i] = &xlObjects{
+				getDisks:    s.GetDisks(i),
+				getLockers:  s.GetLockers(i),
+				endpoints:   endpoints,
+				nsMutex:     mutex,
+				bp:          bp,
+				mrfUploadCh: make(chan partialUpload, 10000),
+			}
+		}(i)
+	}
 
-		// Initialize xl objects for a given set.
-		s.sets[i] = &xlObjects{
-			getDisks:    s.GetDisks(i),
-			getLockers:  s.GetLockers(i),
-			endpoints:   endpoints,
-			nsMutex:     mutex,
-			bp:          bp,
-			mrfUploadCh: make(chan partialUpload, 10000),
-		}
+	wg.Wait()
 
+	for i := 0; i < s.setCount; i++ {
 		go s.sets[i].cleanupStaleMultipartUploads(context.Background(),
 			GlobalMultipartCleanupInterval, GlobalMultipartExpiry, GlobalServiceDoneCh)
-	}
-
-	// Rely on endpoints list to initialize, init lockers.
-	for i := 0; i < s.setCount; i++ {
 		for j := 0; j < s.drivesPerSet; j++ {
-			s.xlLockers[i][j] = newLockAPI(s.endpoints[i*s.drivesPerSet+j])
+			if s.xlDisks[i][j] != nil && s.xlDisks[i][j].IsOnline() {
+				onlineDisks++
+			}
 		}
 	}
 
-	// Connect disks right away, but wait until we have `format.json` quorum.
-	s.connectDisksWithQuorum()
+	if onlineDisks < len(s.endpoints)/2 {
+		// Connect disks right away, but wait until we have `format.json` quorum.
+		s.connectDisksWithQuorum()
```

Comment:
you can remove connectDisksWithQuorum() as it's no longer needed

Reply:
Makes senss